File: examples/poisson-regression.cc
#include "cnn/grad-check.h"
#include "cnn/nodes.h"
#include "cnn/cnn.h"
#include "cnn/training.h"
#include "cnn/timing.h"
#include "cnn/rnn.h"
#include "cnn/gru.h"
#include "cnn/deep-lstm.h"
#include "cnn/lstm.h"
#include "cnn/dict.h"
# include "cnn/expr.h"

#include <iostream>
#include <fstream>
#include <sstream>

#include <boost/archive/text_iarchive.hpp>
#include <boost/archive/text_oarchive.hpp>

using namespace std;
using namespace cnn;

unsigned LAYERS = 2;
unsigned INPUT_DIM = 16;  //256
unsigned HIDDEN_DIM = 32;  // 1024
unsigned VOCAB_SIZE = 0;

cnn::Dict d;
int kSOS;
int kEOS;

template <class Builder>
struct RNNLengthPredictor {
  LookupParameters* p_c;
  Parameters* p_R;
  Parameters* p_bias;
  Builder builder;
  explicit RNNLengthPredictor(Model& model) : builder(LAYERS, INPUT_DIM, HIDDEN_DIM, &model) {
    p_c = model.add_lookup_parameters(VOCAB_SIZE, {INPUT_DIM}); 
    p_R = model.add_parameters({1, HIDDEN_DIM});
    p_bias = model.add_parameters({1});
  }

  // return Expression of total loss
  Expression BuildLMGraph(const vector<int>& sent, unsigned len, ComputationGraph& cg, bool flag = false) {
    const unsigned slen = sent.size() - 1;
    builder.new_graph(cg);  // reset RNN builder for new graph
    builder.start_new_sequence();
    Expression R = parameter(cg, p_R);
    Expression bias = parameter(cg, p_bias);
    vector<Expression> errs;
    for (unsigned t = 0; t < slen; ++t) {
      Expression i_x_t = lookup(cg, p_c, sent[t]);
      // y_t = RNN(x_t)
      builder.add_input(i_x_t);
    }
    Expression pred = affine_transform({bias, R, builder.back()});
    if (flag) {
      unsigned x = exp(as_scalar(cg.incremental_forward()));
      cerr << "PRED=" << x << " TRUE=" << len << "\t(DIFF=" << ((int)x - (int)len) << ")" << endl;
    }
    return poisson_loss(pred, len);
  }
};

int main(int argc, char** argv) {
  cnn::Initialize(argc, argv);
  if (argc != 3 && argc != 4) {
    cerr << "Usage: " << argv[0] << " corpus.txt dev.txt [model.params]\n";
    return 1;
  }
  kSOS = d.Convert("<s>");
  kEOS = d.Convert("</s>");
  vector<pair<vector<int>,unsigned>> training, dev;
  string line;
  int tlc = 0;
  int ttoks = 0;
  cerr << "Reading training data from " << argv[1] << "...\n";
  {
    Dict td;
    ifstream in(argv[1]);
    assert(in);
    while(getline(in, line)) {
      ++tlc;
      vector<int> x, ty;
      ReadSentencePair(line, &x, &d, &ty, &td);
      assert(ty.size() == 1);
      const string& v = td.Convert(ty[0]);
      for(auto c : v) { assert(c >= '0' && c <= '9'); }
      unsigned y = atoi(v.c_str());
      training.push_back(make_pair(x,y));
      ttoks += training.back().first.size();
      if (training.back().first.front() != kSOS && training.back().first.back() != kEOS) {
        cerr << "Training sentence in " << argv[1] << ":" << tlc << " didn't start or end with <s>, </s>\n";
        abort();
      }
    }
    cerr << tlc << " lines, " << ttoks << " tokens, " << d.size() << " types\n";
  }
  d.Freeze(); // no new word types allowed
  VOCAB_SIZE = d.size();

  int dlc = 0;
  int dtoks = 0;
  cerr << "Reading dev data from " << argv[2] << "...\n";
  {
    ifstream in(argv[2]);
    assert(in);
    Dict td;
    while(getline(in, line)) {
      ++dlc;
      vector<int> x, ty;
      ReadSentencePair(line, &x, &d, &ty, &td);
      assert(ty.size() == 1);
      const string& v = td.Convert(ty[0]);
      for(auto c : v) { assert(c >= '0' && c <= '9'); }
      unsigned y = atoi(v.c_str());
      dev.push_back(make_pair(x,y));
      dtoks += dev.back().first.size();
      if (dev.back().first.front() != kSOS && dev.back().first.back() != kEOS) {
        cerr << "Dev sentence in " << argv[2] << ":" << tlc << " didn't start or end with <s>, </s>\n";
        abort();
      }
    }
    cerr << dlc << " lines, " << dtoks << " tokens\n";
  }
  ostringstream os;
  os << "lm"
     << '_' << LAYERS
     << '_' << INPUT_DIM
     << '_' << HIDDEN_DIM
     << "-pid" << getpid() << ".params";
  const string fname = os.str();
  cerr << "Parameters will be written to: " << fname << endl;
  double best = 9e+99;

  Model model;
  Trainer* sgd = nullptr;
  sgd = new SimpleSGDTrainer(&model);

  RNNLengthPredictor<LSTMBuilder> lm(model);
  if (argc == 4) {
    string fname = argv[3];
    ifstream in(fname);
    boost::archive::text_iarchive ia(in);
    ia >> model;
  }

  unsigned report_every_i = 50;
  unsigned dev_every_i_reports = 20;
  unsigned si = training.size();
  vector<unsigned> order(training.size());
  for (unsigned i = 0; i < order.size(); ++i) order[i] = i;
  bool first = true;
  int report = 0;
  unsigned lines = 0;
  while(1) {
    Timer iteration("completed in");
    double loss = 0;
    unsigned chars = 0;
    for (unsigned i = 0; i < report_every_i; ++i) {
      if (si == training.size()) {
        si = 0;
        if (first) { first = false; } else { sgd->update_epoch(); }
        cerr << "**SHUFFLE\n";
        shuffle(order.begin(), order.end(), *rndeng);
      }

      // build graph for this instance
      ComputationGraph cg;
      auto& sent = training[order[si]];
      ++si;
      lm.BuildLMGraph(sent.first, sent.second, cg);
      loss += as_scalar(cg.forward());
      cg.backward();
      sgd->update();
      ++lines;
      ++chars;
    }
    sgd->status();
    cerr << " E = " << (loss / chars) << " ppl=" << exp(loss / chars) << ' ';

    // show score on dev data?
    report++;
    if (report % dev_every_i_reports == 0) {
      double dloss = 0;
      int dchars = 0;
      for (auto& sent : dev) {
        ComputationGraph cg;
        lm.BuildLMGraph(sent.first, sent.second, cg, true);
        dloss += as_scalar(cg.forward());
        dchars++;
      }
      if (dloss < best) {
        best = dloss;
        ofstream out(fname);
        boost::archive::text_oarchive oa(out);
        oa << model;
      }
      cerr << "\n***DEV [epoch=" << (lines / (double)training.size()) << "] E = " << (dloss / dchars) << " ppl=" << exp(dloss / dchars) << ' ';
    }
  }
  delete sgd;
}

File: swig/src/main/scala/edu/cmu/dynet/examples/PoissonRegression.scala
package edu.cmu.dynet.examples

import edu.cmu.dynet._
import edu.cmu.dynet.dynet_swig._

import DynetScalaHelpers._

import scala.language.implicitConversions

import java.nio.file.Paths

object PoissonRegression {

  val LAYERS = 2
  val INPUT_DIM = 16
  val HIDDEN_DIM = 32
  var VOCAB_SIZE = 0

  class RNNLengthPredictor(model: Model) {
    val builder = new LSTMBuilder(LAYERS, INPUT_DIM, HIDDEN_DIM, model)
    val p_c = model.add_lookup_parameters(VOCAB_SIZE, dim(INPUT_DIM))
    val p_R = model.add_parameters(dim(1, HIDDEN_DIM))
    val p_bias = model.add_parameters(dim(1))

    def buildLMGraph(sent: IntVector,
                     len: Int,
                     cg: ComputationGraph,
                     flag: Boolean=false): Expression = {
      val slen = (sent.size - 1).toInt
      builder.new_graph(cg)
      builder.start_new_sequence()

      val R = parameter(cg, p_R)
      val bias = parameter(cg, p_bias)

      for (t <- 0 until slen) {
        val i_x_t = lookup(cg, p_c, sent.get(t))
        builder.add_input(i_x_t)
      }

      val pred = affine_transform(Seq(bias, R, builder.back))

      if (flag) {
        val x = math.exp(cg.incremental_forward(pred).toFloat)
        println(s"PRED = ${x} TRUE = ${len} DIFF = ${x - len}")
      }

      poisson_loss(pred, len)
    }
  }

  def main(args: Array[String]) {
    initialize(new DynetParams)

    val userDir = System.getProperty("user.dir")

    val CORPUS_FILE = Paths.get(userDir, "../examples/cpp/example-data/train-poi.txt").toString
    val DEV_FILE = Paths.get(userDir, "../examples/cpp/example-data/dev-poi.txt").toString

    val d = new WordDict()
    val kSOS = d.convert("<s>")
    val kEOS = d.convert("</s>")

    val training = new scala.collection.mutable.ArrayBuffer[(IntVector, Int)]
    val dev = new scala.collection.mutable.ArrayBuffer[(IntVector, Int)]

    var tlc = 0
    var ttoks = 0

    {
      val td = new WordDict()
      for (line <- scala.io.Source.fromFile(CORPUS_FILE).getLines) {
        tlc += 1

        val (x, y) = WordDict.read_sentence_pair(line, d, td)

        training.append((x, y))

        ttoks += x.size().toInt

        if (x.get(0) != kSOS && x.get(x.size.toInt - 1) != kEOS) {
          throw new RuntimeException("bad sentence")
        }
      }
      println(s"${tlc} lines ${ttoks} tokens ${d.size} types")
    }

    d.freeze()
    VOCAB_SIZE = d.size

    var dlc = 0
    var dtoks = 0

    {
      val td = new WordDict
      for (line <- scala.io.Source.fromFile(DEV_FILE).getLines) {
        dlc += 1

        val (x, y) = WordDict.read_sentence_pair(line, d, td)

        dev.append((x, y))

        dtoks += x.size().toInt

        if (x.get(0) != kSOS && x.get(x.size.toInt - 1) != kEOS) {
          throw new RuntimeException("bad sentence")
        }
      }
      println(s"${dlc} lines ${dtoks} tokens")
    }

    var best = Float.MaxValue

    val model = new Model()
    val sgd = new SimpleSGDTrainer(model)

    val lm = new RNNLengthPredictor(model)

    val report_every_i = 50
    val dev_every_i_reports = 20
    var si = training.size
    val order = new IntVector((0 until training.size))

    var first = true
    var report = 0
    var lines = 0

    val cg = ComputationGraph.getNew

    while (true) {
      var loss = 0.0f
      var chars = 0
      for (i <- 0 until report_every_i) {
        if (si == training.size) {
          si = 0
          if (first) {
            first = false
          } else {
            sgd.update_epoch()
          }

          shuffle(order)
        }

        // build graph for this instance

        // the cg.clear is IMPORTANT!
        cg.clear()

        val (tokens, count) = training(order.get(si))

        si += 1

        val loss_expr = lm.buildLMGraph(tokens, count, cg)
        loss += cg.forward(loss_expr).toFloat
        cg.backward(loss_expr)
        sgd.update()
        lines += 1
        chars += 1
      }
      sgd.status()
      println(s"E = ${loss / chars.toFloat} ppl = ${math.exp(loss / chars.toFloat)}")

      report += 1
      if (report % dev_every_i_reports == 0) {
        var dloss = 0.0f
        var dchars = 0

        for ((tokens, count) <- dev) {
          val loss_expr = lm.buildLMGraph(tokens, count, cg, true)
          dloss += cg.forward(loss_expr).toFloat
          dchars += 1
        }

        if (dloss < best) {
          best = dloss
          println(s"new best: ${best}")
        }

        println(s"DEV [epoch = ${lines / training.size.toFloat}] " +
                s"E = ${dloss / dchars.toFloat} " +
                s"ppl = ${math.exp(dloss / dchars.toFloat)}")
      }
    }
  }
}
